### PR TITLE
http3: further cleanup and extension removal

### PR DESCRIPTION
**Pull request by mattklein123**
Fixes https://github.com/envoyproxy/envoy/issues/12829

Risk Level: Low
Testing: Existing/new tests
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

**Comment by mattklein123** [comment]
Opening as a draft to check CI. I need to add more test coverage of the QUIC disabled case to make sure we do sufficient configuration error checking and can't hit any of the not reached guards.

File: source/common/quic/client_connection_factory_impl.cc [hunk 1]
```diff
@@ -24,11 +24,21 @@ PersistentQuicInfoImpl::PersistentQuicInfoImpl(
           std::make_unique<quic::QuicCryptoClientConfig>(std::make_unique<EnvoyQuicProofVerifier>(
               stats_scope, getConfig(transport_socket_factory), time_source))) {}
 
+namespace {
+// TODO(mattklein123): This is mutable static info that is required for the QUICHE code. This was
+// preexisting but should be removed as we shouldn't be using global mutable static data.
```
Comment (mattklein123) [inline source/common/quic/client_connection_factory_impl.cc:29]: cc @danzh2010 @alyssawilk for comment on this one.

**Comment by mattklein123** [comment]
@alyssawilk @danzh2010 I think this should be ready for review.

File: test/common/quic/BUILD [hunk 2]
```diff
@@ -281,3 +281,26 @@ envoy_cc_test_library(
         "@com_googlesource_quiche//:quic_test_tools_qpack_qpack_encoder_test_utils_lib",
     ],
 )
+
+envoy_cc_test(
```
Comment (mattklein123) [inline test/common/quic/BUILD:285]: @alyssawilk I think you accidentally deleted this here: https://github.com/envoyproxy/envoy/pull/15540/files#diff-679677aa5d85fe778bee44c14d7a92b9d47b29f4a779557d7307f8ab39f074a6L12. Let me know if that's not the cae?

File: source/common/http/http3/conn_pool.cc [hunk 3]
```diff
@@ -7,12 +7,17 @@
...
-#include "common/http/http3/well_known_names.h"
 #include "common/http/utility.h"
 #include "common/network/address_impl.h"
 #include "common/network/utility.h"
 #include "common/runtime/runtime_features.h"
 
+#ifdef ENVOY_ENABLE_QUIC
+#include "common/quic/client_connection_factory_impl.h"
+#else
+#error "http3 conn pool should not be built with QUIC disabled"
```
Comment (alyssawilk) [inline source/common/http/http3/conn_pool.cc:18]: Per moderation's issue with quic header files somehow getting included with the !http3 build, I wonder if we should have a macro for this and include it in all the quic headers right under pragma once.  WDYT?

File: source/common/quic/client_connection_factory_impl.cc (hunk 1 above)
Comment (alyssawilk) [reply source/common/quic/client_connection_factory_impl.cc:29]: you're welcome to TODO this my way, alternately we can probably stick both in PersistentQuicInfo, or make push promise index there and make quic_config_ const at which point it's legal I think?

File: source/extensions/extensions_build_config.bzl [hunk 4]
```diff
@@ -262,7 +262,7 @@ EXTENSIONS = {
     # HTTP header formatters
     #
 
-    "envoy.http.stateful_header_formatters.preserve_case":       "//source/extensions/http/header_formatters/preserve_case:preserve_case_formatter"
+    "envoy.http.stateful_header_formatters.preserve_case":       "//source/extensions/http/header_formatters/preserve_case:preserve_case_formatter",
```
Comment (alyssawilk) [inline source/extensions/extensions_build_config.bzl:265]: revert?

File: test/common/quic/BUILD (hunk 2 above)
Comment (alyssawilk) [reply test/common/quic/BUILD:285]: deleted on purpose: I moved the test into  test/common/upstream/upstream_impl_test.cc to address the TODO


#ifdef ENVOY_ENABLE_QUIC
TEST_F(ClusterInfoImplTest, Http3) {

...

**Review by alyssawilk — COMMENTED** [review]
So.  Much.  Cleaner.

Honestly I would tend to stick with some of the factories over proliferation of #defines but that's style and I'm good either way.

File: source/common/http/http3/conn_pool.cc (hunk 3 above)
Comment (mattklein123) [reply source/common/http/http3/conn_pool.cc:18]: The main issue with that is right now we actually still run the QUIC tests when QUIC is compiled out. Let me see if I can fix this as I agree it would be cleaner.

File: source/common/quic/client_connection_factory_impl.cc (hunk 1 above)
Comment (mattklein123) [reply source/common/quic/client_connection_factory_impl.cc:29]: OK let me take another look.

**Comment by mattklein123** [comment]
> Honestly I would tend to stick with some of the factories over proliferation of #defines but that's style and I'm good either way.

I can take a look at stubs instead, which I may have to do if we want to fully compile everything out even for tests. I will take another look.

File: source/common/quic/client_connection_factory_impl.cc (hunk 1 above)
Comment (alyssawilk) [reply source/common/quic/client_connection_factory_impl.cc:29]: ah, and having checked in, bence is starting to remove server push upstream so hopefully push promise index will go away soonish regardless (plus remove a bunch of uncovered stubs)

Comment (mattklein123) [reply source/common/quic/client_connection_factory_impl.cc:29]: OK if this is going to change soon I will probably just punt since this is no different from what it was before.

**Comment by mattklein123** [comment]
@PiotrSikora the last commit should fix the fips build. cc @keith thanks for working through that with me.

@alyssawilk this whole situation is honestly a total mess and I think we clean it up with a new feature in bazel 4.0 which supports targets being "compatible with" flags. I would suggest we land this, upgrade to 4.0, and then we can clean up further with actually fully blocking the targets if disabled?cc @lizan

**Comment by PiotrSikora** [comment]
> @PiotrSikora the last commit should fix the fips build. cc @keith thanks for working through that with me.

Thanks! I can confirm it fixes the issue.

File: source/common/http/http3/conn_pool.cc (hunk 3 above)
Comment (alyssawilk) [reply source/common/http/http3/conn_pool.cc:18]: another thought along these lines,what do we think of a macro
QUIC_INCLUDE( "common/quic/client_connection_factory_impl.h")
which translates into this?

having the ifndef inline is more clear, but having a macro does seem easier, and we could optionally late format check nothing in quic is straight up included.

File: source/common/quic/client_connection_factory_impl.cc (hunk 1 above)
Comment (alyssawilk) [reply source/common/quic/client_connection_factory_impl.cc:29]: yeah, if you don't mind changing the TODO to the quic upstream bug that'd be great (since I'm tracking those).

File: test/common/http/BUILD [hunk 5]
```diff
@@ -415,10 +415,11 @@ envoy_cc_test(
     ],
 )
 
+# TODO(mattklein123): Use compatible with when we move to Bazel 4.0.
 envoy_cc_test(
     name = "conn_pool_grid_test",
-    srcs = ["conn_pool_grid_test.cc"],
-    deps = [
+    srcs = envoy_select_enable_http3(["conn_pool_grid_test.cc"]),
```
Comment (alyssawilk) [inline test/common/http/BUILD:421]: I think this one can stay as-is.  The test should be intrumented to run the HTTP/3 checks when HTTP/3 is on, and just test the grid (eventually the ipv4/ipv6 parts) when http/3 is off.

File: bazel/BUILD [hunk 6]
```diff
@@ -299,6 +299,14 @@ config_setting(
     },
 )
 
+selects.config_setting_group(
```
Comment (alyssawilk) [inline bazel/BUILD:302]: Hey @lizan can you do a pass of just the bazel magic?

**Review by alyssawilk — COMMENTED** [review]
Just when we thought it couldn't get any harier :-P

question - do you think we should have the compile options build do one pass with 
!quic
and another with 
!fips
to futureproof against that sort of regression?  I'm inlined to think it's worth it but we can always try with this as-is and revisit if there are future breakages

File: source/common/quic/client_connection_factory_impl.cc (hunk 1 above)
Comment (danzh2010) [reply source/common/quic/client_connection_factory_impl.cc:29]: Why do we need to remove QuicClientConnectionFactoryImpl? It makes sense to remove the abstraction, but I think we need to stick some quiche related stuff in connection pool any way, why not just wrap everything in this factory, including config_? And this factory which depends on QUICHE can be hide behind ENVOY_ENABLE_QUIC

Comment (mattklein123) [reply source/common/quic/client_connection_factory_impl.cc:29]: > Why do we need to remove QuicClientConnectionFactoryImpl? 

Because the indirection (and lookup cost) isn't needed. Essentially, the factory is a global object, so what I have here is exactly the same as what we had before. My preference still is to just leave this as it is the same and we can sort it out in a future change? This PR is quite complicated and I'm trying to not complicate it further. :)

File: source/common/http/http3/conn_pool.cc (hunk 3 above)
Comment (mattklein123) [reply source/common/http/http3/conn_pool.cc:18]: Sure I can do this. Like I said elsewhere, there is a much better way of dealing with this once we have Bazel 4.0, which will drop all of the QUIC rules entirely so it simply won't build. I'm happy to do the macro though to make it a bit visually cleaner.

File: test/common/http/BUILD (hunk 5 above)
Comment (mattklein123) [reply test/common/http/BUILD:421]: Without this it didn't compile, so I suspect that at runtime the old code would have failed due to missing factories if compiled out? Can I live this and you can decide how you want the code to work without H3 compiled in since it's new code?

File: bazel/BUILD (hunk 6 above)
Comment (mattklein123) [reply bazel/BUILD:302]: cc @keith also

File: test/common/http/BUILD (hunk 5 above)
Comment (alyssawilk) [reply test/common/http/BUILD:421]: Do you know which compile it failed?  Both the quic includes and libraries were already selected.
but yeah, feel free to toss in a TODO for me

File: source/common/quic/client_connection_factory_impl.cc (hunk 1 above)
Comment (danzh2010) [reply source/common/quic/client_connection_factory_impl.cc:29]: A TODO to move config_ into PersistentQuicInfo is fine. Or even better TODO to move the creation logic into PersistentQuicInfo.

Comment (mattklein123) [reply source/common/quic/client_connection_factory_impl.cc:29]: OK will do.

File: bazel/BUILD (hunk 6 above)
Comment (keith) [reply bazel/BUILD:302]: As long as this works it's definitely fine, I'm impressed it does though, we should separately change the apple config to use this instead, looking at the skylib impl it's very hard to parse, but I believe it's doing that same hack https://github.com/bazelbuild/bazel-skylib/blob/f80bc733d4b9f83d427ce3442be2e07427b2cc8d/lib/selects.bzl#L145-L186

Comment (mattklein123) [reply bazel/BUILD:302]: Believe it or not, I found this trick by googling, where I found the bazel issue you opened about the apple hack, where the person commented it could be done with this thing, but it wasn't merged yet. Now it's merged. :)

File: source/common/http/http3/conn_pool.cc (hunk 3 above)
Comment (mattklein123) [reply source/common/http/http3/conn_pool.cc:18]: OK I tried and failed on this. I don't think there is any way to define a macro that does include or error. We can circle back on this and see if we can make it better later.

**Review by alyssawilk — APPROVED** [review]

File: bazel/BUILD (hunk 6 above)
Comment (keith) [reply bazel/BUILD:302]: Amazing, I forgot this, here's the update for the other ones that can benefit from this https://github.com/envoyproxy/envoy/pull/15797

**Comment by bencebeky** [comment]
The change to HttpIntegrationTest::makeClientConnectionWithOptions() in test/integration/http_integration.cc breaks compilation on certain embedders with the following error message:
third_party/envoy/src/test/integration/http_integration.cc:239:1: error: non-void function does not return a value in all control paths [-Werror,-Wreturn-type]

This is part of presubmit and thus blocking submission of any CL for our project.  Please fix.  (Unfortunately I do not have my Envoy checkout any more, otherwise I would fix it myself.)

**Comment by alyssawilk** [comment]
I'll take a look
